File: backend/internal/domain/network.go
package domain

import (
	"context"
	"time"

	"github.com/google/uuid"
)

// NetworkRole mirrors the Postgres network_role enum.
type NetworkRole string

const (
	NetworkRoleOwner  NetworkRole = "owner"
	NetworkRoleAdmin  NetworkRole = "admin"
	NetworkRoleMember NetworkRole = "member"
)

// Network is a virtual LAN / mesh network with a CIDR address space.
type Network struct {
	ID                  uuid.UUID
	Name                string
	Description         string
	CIDR                string
	CIDRv6              *string
	DNSServers          []string
	OwnerID             uuid.UUID
	InviteCode          string
	InviteCodeExpiresAt *time.Time
	InviteEnabled       bool
	AllowBroadcast      bool
	CreatedAt           time.Time
	UpdatedAt           time.Time

	// Populated by joined queries, not persisted directly on this table.
	MemberCount int
	DeviceCount int
	// CallerRole is the requesting user's role in this network, populated
	// by queries that join against network_members for "my networks" views.
	CallerRole NetworkRole
}

// NetworkMember links a user to a network with a role.
type NetworkMember struct {
	ID        uuid.UUID
	NetworkID uuid.UUID
	UserID    uuid.UUID
	Role      NetworkRole
	JoinedAt  time.Time

	// Populated by joined queries.
	Email       string
	DisplayName string
}

// NetworkRepository persists Network aggregates.
type NetworkRepository interface {
	Create(ctx context.Context, n *Network) error
	GetByID(ctx context.Context, id uuid.UUID) (*Network, error)
	GetByInviteCode(ctx context.Context, code string) (*Network, error)
	ListForUser(ctx context.Context, userID uuid.UUID) ([]*Network, error)
	Update(ctx context.Context, n *Network) error
	Delete(ctx context.Context, id uuid.UUID) error
	RotateInviteCode(ctx context.Context, id uuid.UUID, newCode string, expiresAt *time.Time) error
	CountMembers(ctx context.Context, networkID uuid.UUID) (int, error)
	CountDevices(ctx context.Context, networkID uuid.UUID) (int, error)
}

// NetworkMemberRepository persists network membership rows.
type NetworkMemberRepository interface {
	Create(ctx context.Context, m *NetworkMember) error
	Get(ctx context.Context, networkID, userID uuid.UUID) (*NetworkMember, error)
	List(ctx context.Context, networkID uuid.UUID) ([]*NetworkMember, error)
	ListNetworkIDsForUser(ctx context.Context, userID uuid.UUID) ([]uuid.UUID, error)
	UpdateRole(ctx context.Context, networkID, userID uuid.UUID, role NetworkRole) error
	Delete(ctx context.Context, networkID, userID uuid.UUID) error
	CountByRole(ctx context.Context, networkID uuid.UUID, role NetworkRole) (int, error)
}

File: mobile/lib/core/models/network.dart
/// Mirrors the `Network` schema in api/openapi.yaml.
class Network {
  final String id;
  final String name;
  final String? description;
  final String cidr;
  final List<String> dnsServers;
  final String role; // owner | admin | member
  final int memberCount;
  final int deviceCount;
  final String? inviteCode;
  final DateTime? createdAt;

  const Network({
    required this.id,
    required this.name,
    this.description,
    required this.cidr,
    this.dnsServers = const [],
    required this.role,
    this.memberCount = 0,
    this.deviceCount = 0,
    this.inviteCode,
    this.createdAt,
  });

  factory Network.fromJson(Map<String, dynamic> json) {
    return Network(
      id: json['id'] as String? ?? '',
      name: json['name'] as String? ?? '',
      description: json['description'] as String?,
      cidr: json['cidr'] as String? ?? '',
      dnsServers: (json['dnsServers'] as List<dynamic>? ?? const [])
          .map((e) => e.toString())
          .toList(),
      role: json['role'] as String? ?? 'member',
      memberCount: json['memberCount'] as int? ?? 0,
      deviceCount: json['deviceCount'] as int? ?? 0,
      inviteCode: json['inviteCode'] as String?,
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        if (description != null) 'description': description,
        'cidr': cidr,
        'dnsServers': dnsServers,
        'role': role,
        'memberCount': memberCount,
        'deviceCount': deviceCount,
        if (inviteCode != null) 'inviteCode': inviteCode,
        if (createdAt != null) 'createdAt': createdAt!.toIso8601String(),
      };

  Network copyWith({
    String? name,
    String? description,
    String? cidr,
    List<String>? dnsServers,
    String? inviteCode,
  }) {
    return Network(
      id: id,
      name: name ?? this.name,
      description: description ?? this.description,
      cidr: cidr ?? this.cidr,
      dnsServers: dnsServers ?? this.dnsServers,
      role: role,
      memberCount: memberCount,
      deviceCount: deviceCount,
      inviteCode: inviteCode ?? this.inviteCode,
      createdAt: createdAt,
    );
  }

  bool get canManage => role == 'owner' || role == 'admin';
}
